File: DSA/week12/quicksort.py
from random import randint
def quicksort(arr, low, high):
    if low >= high:
        return
    
    s = low
    e = high
    mid = s + (e-s)//2  
    pivot = arr[mid]
    while s <= e:
        while arr[s] < pivot:
            s+=1
        while arr[e] > pivot:
            e-=1
        
        if (s<=e):
            arr[s], arr[e] = arr[e], arr[s]
            s+=1
            e-=1
    quicksort(arr, low, e)
    quicksort(arr, s, high)


n = [randint(0, 100) for i in range(20)]
arr = [0, 2, 1, 9, 5]
print(arr)
quicksort(arr, 0, len(arr)-1)
print(arr)

File: DSA/week12/quicksort.ts
function quicksort(nums: number[], low: number, high: number) {
    //base case.
    if (low >= high) {
        return;
    }

    let s = low;
    let e = high;
    let m = s + Math.floor((e-s)/2) //middle
    let pivot = nums[m];

    while (s <= e) {
        //if it's sorted, it will not swap
        while (nums[s] < pivot) {
            s++;
        }
        while (nums[e] > pivot) {
            e--;
        }

        if (s<=e) {
            //swap
            let temp = nums[s]
            nums[s] = nums[e]
            nums[e] = temp
            s++;
            e--;
        }
    }

    //pivot is now at current index, sort the 2 halves
    quicksort(nums, low, e);
    quicksort(nums, s, high);

}

let numbers: number[] = [2, 3, 8, 1, 89, 54, 5, 0]
quicksort(numbers, 0, numbers.length-1)
console.log(numbers)